### PR TITLE
Secured media and public media

### PR DESCRIPTION
**Pull request by taylor-work**
Issue: *Strapi has one kind of media asset, public. We need to provide public and secured media assets. If a media file is prepended with `public_` it can be accessed without authentication. If this prefix is missing authentication is required for the media file.*

Uses https://github.com/flexion/strapi-provider-upload-aws-s3-auth . Assumes a private s3 bucket ACL's. The upload provider is a fork of the s3 provider and returns urls that this plugin can intercept and use.

### TODO
 - [x] Need endpoint to set cookie.
 - [x] Need middleware to move cookie into authorization header.
 - [x] Need to sort out FQDN's + third party cookies.

### CODE REVIEW

File: cms/plugins/upload-auth/controllers/upload-auth.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+module.exports = {
+
+  /**
+   * Default action.
+   *
+   * @return {Object}
+   */
+
+  index: async (ctx) => {
```
Comment (taylor-work) [inline cms/plugins/upload-auth/controllers/upload-auth.js:17]: Majority of the work is here.

File: cms/plugins/upload-auth/controllers/upload-auth.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    const requestedFile = ctx.captures[0];
+    
+    if (ctx.isAuthenticated() || publicAssetMatcher.test(requestedFile)) {
+      const data = await s3.getObject({
+        Bucket: bucket,
+        Key: requestedFile,
+      }).promise();
+
+      ctx.type = data.ContentType;
+      return ctx.send(data.Body);
```
Comment (danielnaab) [inline cms/plugins/upload-auth/controllers/upload-auth.js:30]: Can we attach the s3 data directly to the response body, to stream the s3 object back? `response.body=` here seems to indicate we can https://koajs.com/#response

Comment (taylor-work) [reply cms/plugins/upload-auth/controllers/upload-auth.js:30]: I'm not a koa guru at all. I searched for the equivalent of `res.sendFile` and found several closed issues and pull requests. That gave me the impression that's how koa is designed to be used. If you can find a better API I'm game.

Comment (taylor-work) [reply cms/plugins/upload-auth/controllers/upload-auth.js:30]: Also we're dealing with the API's aws sdk expose to us, `data.Body` is already a buffer so koa may be doing the right thing.

**Review by danielnaab — APPROVED** [review]

**Pull request merged**